python/ray/data/_internal/execution/util.py

```diff
@@ -75,8 +77,12 @@ def __repr__(self):
             return super().__repr__()
 
         def __call__(self, *args, **kwargs):
-            # ThreadPoolExecutor will reuse the same thread for every submit call.
-            future = self.thread_pool_executor.submit(super().__call__, *args, **kwargs)
-            return future.result()
+            if single_threaded:
+                # ThreadPoolExecutor will reuse the same thread for every submit call.
+                future = self.thread_pool_executor.submit(
+                    super().__call__, *args, **kwargs
+                )
+                return future.result()
+            return super().__call__(*args, **kwargs)
```

Comment:
You can skip the whole wrapper in that case

Reply:
Let's also make it clear that this is a single-threaded wrapper used to restrict concurrency by default